Sources/MongoMobile/MongoMobileLogger.swift

```diff
@@ -0,0 +1,99 @@
+/// If the user provided a logger when initializing the library, executes the 
+/// `onMessage` callback for that logger.
+internal func mongo_mobile_log_callback(userDataPtr: UnsafeMutableRawPointer?,
+                                        messagePtr: UnsafePointer<Int8>?,
+                                        componentPtr: UnsafePointer<Int8>?,
+                                        contextPtr: UnsafePointer<Int8>?,
+                                        severity: Int32) {
+    guard let logger = MongoMobile.logger else {
+        return
+    }
+
+    let message = String(from: messagePtr)
+    let component = String(from: componentPtr)
+    let context = String(from: contextPtr)
+    // swiftlint:disable:next force_unwrapping - server should always gives a valid raw value.
+    let severity = LogSeverity(rawValue: severity)!
+    logger.onMessage(message: message, component: component, context: context, severity: severity)
+}
+
+extension String {
+    /// Returns a string created from the given pointer, or an empty
+    /// string if the pointer is nil.
+    fileprivate init(from ptr: UnsafePointer<Int8>?) {
+        if let ptr = ptr {
+            self = String(cString: ptr)
+        } else {
+            self = ""
+        }
+    }
+}
+
+/// Defines the requirements for a type that can be used as a logger
+/// for the embedded library.
+public protocol MongoMobileLogger {
+    /**
+    * A callback that will be executed when a new message is produced
+    * by an embedded server.
+    *
+    * - Parameters:
+    *   - message: The text of the log message.
+    *   - component: Functional categorization of the message, e.g. "query" or "network".
+    *   - context: Describes the context in which this message occurred, e.g. "initandlisten".
+    *   - severity: The severity level associated with this message.
+    * 
+    * - SeeAlso: https://docs.mongodb.com/manual/reference/log-messages/
+    */
+    func onMessage(message: String,
```

Comment:
a downside of this API is that it requires us to always convert all the associated strings into Swift types, even if the user ends up not caring about them and ignores the message due to its log level.
I got to thinking about this reading over the [server work group logging API proposal](https://forums.swift.org/t/feedback-server-logging-api-with-revisions/19375). 

they define similar methods on a `Logger` type, but each of them looks like this:
```swift
public func log(level: Logging.Level, message: @autoclosure () -> String, metadata: @autoclosure () -> Logging.Metadata? = nil, error: Error? = nil)
public func debug(_ message: @autoclosure () -> String, metadata: @autoclosure () -> Logging.Metadata? = nil)
```

Their justification: "Both `message` and `metadata` are `@autoclosure` parameters which means that if the message does not end up being logged (because it's below the currently configured log level) no unnecessary processing is done rendering the `String` or creating the metadata information."

so we could change this method to look like:
```swift
func onMessage(message: @autoclosure () -> String,
               component: @autoclosure () -> String,
               context: @autoclosure () -> String,
               severity: LogSeverity)
```

(I suppose `severity` could even be an autoclosure too.)

Then when we call it above in `mongo_mobile_log_callback` we would directly pass in the `String` initializer calls: `onMessage(message: String(from: messagePtr), ...)

this definitely seems more performant, but it makes the protocol look a lot more complicated. it also requires the user to access the parameters as method calls, like `component()` instead of `component`, etc.

thoughts?

more generally, this is the first time I've really thought about autoclosures. I bet they would be useful to us in other parts of our codebase where we do expensive copies between C and Swift. good John Sundell post on autoclosures [here](https://www.swiftbysundell.com/posts/using-autoclosure-when-designing-swift-apis)

edit:  I'm not sure what the lifetimes of these pointers are. my guess is they are only valid while `mongo_mobile_log_callback` is still being executed. so we would have to instruct users that they cannot pass the autoclosures outside of their `onMessage` implementation without evaluating them first.

Reply:
all sounds good to me, will wait to hear what @mbroadst thinks before updating.

Reply:
Strongly agree with using autoclosures, and I think you're right that it might behoove us to comb the API to see where else we can benefit from this approach.